### ISSUE TITLE
Examples how to use Uno with gRPC

### ISSUE THREAD

**Issue by ghuntley** [unoplatform/uno#1594]
## What would you like to be added:

Instructions how to use grpc w/Uno (specifically WebAssembly)

## Why is this needed:

Pre-emptive request as there were announcements recently at .NET conf about grpc.

## For which Platform:

- [x] iOS
- [x] Android
- [x] WebAssembly
- [ ] Windows

## Anything else we need to know?

- https://grpc.io/blog/grpc-on-dotnetcore/
- https://github.com/grpc/grpc-web
- https://github.com/grpc/grpc-dotnet

**Comment by jhalak27** [comment]
Hey, I am new here. Could you let me know what needs to de done here?

**Comment by jeromelaban** [comment]
Thanks for taking a stab at it @jhalak27!

A first step would be to try using the commonly known grpc.net samples to see what works, what does not, and post your findings here.

**Comment by MartinZikmund** [comment]
@jhalak27 are you still interested in this one?

**Comment by jeromelaban** [comment]
@inforzip we do not have documentation to use this, as you've seen. If you can provide a repro sample to this issue of what you're trying to do, it'll help.

**Comment by inforzip** [comment]
I apologize for late reply.
I add a project one with the problem that is happening to us with grpc client.
We managed to establish the channel with the grpc server side, but when we need to call the class  new Greeter.GreeterClient(channel) is not present,  we think because the grpc protobuf class generator is not called and does not create the respective classes from the .proto files located in shared project.
Any help is well comming.

**Comment by MarkStega** [comment]
<ItemGroup>
        <Protobuf Include="gRPCClient\codeProto.proto" />
        <Protobuf Include="gRPCClient\persistentStorageProxyProto.proto" />
        <Protobuf Include="gRPCClient\documentProto.proto" />
        <Protobuf Include="gRPCClient\practiceProto.proto" />
        <Protobuf Include="gRPCClient\equipmentProto.proto" />
        <Protobuf Include="gRPCClient\providerProto.proto" />
        <Protobuf Include="gRPCClient\reportDefinitionProto.proto" />
        <Protobuf Include="gRPCClient\reportProto.proto" />
        <Protobuf Include="gRPCClient\userProto.proto" />
    </ItemGroup>

**Comment by inforzip** [comment]
I appreciate all the help given so far.

I have been testing all the suggestions that have come to me, but the problem has not been solved. This leads to frustration due to the time spent and not seeing any light at the end of the tunnel.
I have all the nuget packages installed (grpc.core;grpc.net.client;grpc.tools;google.protobuf).
I manage to create the connection with the grpc server, but when I try to create the client I get the error:
 **The CompileXaml task returned false but did not log an error. testGrpc2.Uwp**
although the classes are created in the \obj folder by google.protobuf.
```
var channel = GrpcChannel.ForAddress("https://localhost:5001", new GrpcChannelOptions
{
    HttpHandler = new System.Net.Http.HttpClientHandler(),
});

//Comment bottom line to cleaar the issue
var client = new Greet.Greeter.GreeterClient(channel);
```
What will be causing the crap of this error. I have already eliminated the \obj and \bin folders and restarted visual studio but the problem remains.
Any clue is welcome.

**Comment by MarkStega** [comment]
@inforzip How about either posting a new zip file or creating a public github repository of your test project?

**Comment by inforzip** [comment]
I've tried everything, please help

Download the project from this link
https://inforzip.com/testGrpc2.zip

**Comment by inforzip** [comment]
I already deleted the \obj folder and recompile the project but the problem remains.
This situation happens in the uno project that I made available but also in a uwp project. 
In console and wpf projects everything seems to work much better

**Comment by jeromelaban** [comment]
@inforzip Thanks for the repro. You'll need to make sure that the proto file is accessible from all "Head" projects (it's only in the UWP project in your sample).

The easiest way is to create a netstandard2.0 library project, place the proto files there, then reference that project in all project heads.

**Comment by inforzip** [comment]
First of all I want to thank you for all the help provided.

With the help of jerome i managed to move forward. Now i can create the channel, create the client, make the request to the grpc server. Everything seemed perfect until i verified that the server receives the client's request, processes it and responds., but the client receives the following error

StatusCode="Cancelled", Detail="No grpc-status found on response".

This is my code:
```
AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http1UnencryptedSupport", true);
var handler = new HttpClientHandler();
handler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
// The error happens on the next line
var channel = GrpcChannel.ForAddress("https://localhost:5001", new GrpcChannelOptions
  {
      HttpHandler = handler,
      Credentials = Grpc.Core.ChannelCredentials.SecureSsl
  });
var client = new Greeter.GreeterClient(channel);
```
Thanks in advance

**Comment by mcNets** [comment]
Hi I am currently using gRPC in a UNO project (Windows + Skia), I can test it with Android emulators but not with IOS, my Apple developer license expired long ago.

Can I get some help with the web head?

**Comment by mcNets** [comment]
Android head issues:

The issue I found with the Android head is related to `grpc_csharp_ext` library. It seems to be a Windows-only library so it has to be somehow compiled specifically for Android.

```
Loaded assembly: /data/data/McNets.UnoGrpc/files/.__override__/System.Text.RegularExpressions.dll [External]
[monodroid-assembly] Shared library 'grpc_csharp_ext' not loaded, p/invoke 'grpcsharp_redirect_log' may fail
[monodroid-assembly] Shared library 'grpc_csharp_ext' not loaded, p/invoke 'grpcsharp_redirect_log' may fail
**System.InvalidOperationException:** 'Failed to load UnoGrpc.MainPage: System.DllNotFoundException: grpc_csharp_ext
```

I'am trying to find a workaround in `grpc` and `Xamarin` repos but I haven't find out yet. It seems that by adding `Grpc.Core` package to the Android head sometimes solve the issue but no here.

You can read about it;

https://github.com/grpc/grpc-dotnet/issues/2126

https://github.com/grpc/grpc/issues/16250

https://github.com/grpc/grpc/issues/16310

**Comment by mcNets** [comment]
Ok, I take this. 

Should I post a blog article?

**Comment by MartinZikmund** [comment]
This should be kept open

**Comment by DevTKSS** [comment]
@MartinZikmund just seen this here, this would be really great to have in the future😍 
is this blocked by the https issue and/or the dotnet/runtime for mulit threading, or something different?

**Comment by jeromelaban** [comment]
@DevTKSS this is not blocked by anything, it's only a matter of assembling the right information.

**Comment by mcNets** [comment]
If I remember correctly, a Discord user sent me a link to an example (repository) using gRPC in a Uno web app a while back. I'll try to recover it.

@MartinZikmund

**Comment by MartinZikmund** [comment]
# Repro generator

This issue was classified as not reproducible with a code sample or test.

**Reason:** This is a documentation request for examples of using gRPC with Uno Platform (specifically WebAssembly). There is no specific broken behavior to reproduce — it is a request for guidance and documentation.

**Comment by mcNets** [comment]
@jeromelaban @MartinZikmund where should this example be added, in the Uno documentation or as an external publication?

**Comment by MartinZikmund** [comment]
> @jeromelaban @MartinZikmund where should this example be added, in the Uno documentation or as an external publication?

I would say docs here + Uno.Samples sample

**Comment by mcNets** [comment]
> > [@jeromelaban](https://github.com/jeromelaban) [@MartinZikmund](https://github.com/MartinZikmund) where should this example be added, in the Uno documentation or as an external publication?
> 
> I would say docs here + Uno.Samples sample

Can I start a grpc server from a Uno sample in the Uno.samples?

Let me build the example and show it to you, and then we can decide how to to add it to the samples.

**Comment by mcNets** [comment]
@MartinZikmund 
Please have a look at this project: https://github.com/mcNets/UnoGrpc

Do you think it can be added to the Uno.samples?

I think it requires a help document on how to configure the different targets.

**Comment by sasakrsmanovic** [comment]
@agneszitte @Jen-Uno  - Pingback on this on Joan's behalf.

**Comment by MartinZikmund** [comment]
@mcNets Yes, I think that can work! Each of the samples in the Uno.Samples repo has a README in its root, so those instructions can be put there

**Comment by mcNets** [comment]
@MartinZikmund please have a look https://github.com/unoplatform/Uno.Samples/pull/913

**Comment by MartinZikmund** [comment]
Merged! https://github.com/unoplatform/Uno.Samples/pull/913
